llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -7283,31 +7283,35 @@ static bool isGuaranteedNotToBeUndefOrPoison(
   // BB1:
   //   CtxI ; V cannot be undef or poison here
   auto *Dominator = DNode->getIDom();
-  while (Dominator) {
-    auto *TI = Dominator->getBlock()->getTerminator();
-
-    Value *Cond = nullptr;
-    if (auto BI = dyn_cast_or_null<BranchInst>(TI)) {
-      if (BI->isConditional())
-        Cond = BI->getCondition();
-    } else if (auto SI = dyn_cast_or_null<SwitchInst>(TI)) {
-      Cond = SI->getCondition();
-    }
+  // This check is purely for compile time reasons: we do not need to do the
+  // IDom walk if we are checking for only undef and the value is not a boolean.
+  if (includesPoison(Kind) || V->getType()->isIntOrIntVectorTy(1))
```

Comment:
Shouldn't this be?
```suggestion
  if (!includesUndef(Kind) || V->getType()->isIntOrIntVectorTy(1))
```
We can also skip it if we're checking for *both* undef an poison, right?

Reply:
That's right, thanks! it should broader the impact of this change as well. Will update the commit message.